### PR TITLE
[chore ] Refactor CoinMarketCap

### PR DESCRIPTION
**Pull request by colmazia**
## Please ensure the following requirements are met before submitting a pull request:

- [x] The pull request is targeted against the correct target branch
- [ ] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [ ] The pull request includes a description of the implementation/work done in detail.
- [x] The pull request includes any appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG.md`
- [x] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the GitHub PR explorer)

### CODE REVIEW

File: bothan-coinmarketcap/src/worker/types.rs [hunk 1]
```diff
@@ -0,0 +1,4 @@
+use tokio::time::Duration;
+
+pub(crate) const DEFAULT_UPDATE_INTERVAL: Duration = Duration::from_secs(60);
+pub(crate) const DEFAULT_UPDATE_SUPPORTED_ASSETS_INTERVAL: Duration = Duration::from_secs(86400);
```
Comment (RogerKSI) [inline bothan-coinmarketcap/src/worker/types.rs:4]: what is this for?

File: bothan-coinmarketcap/src/worker/types.rs [hunk 2]
```diff
@@ -0,0 +1,4 @@
+use tokio::time::Duration;
```
Comment (RogerKSI) [inline bothan-coinmarketcap/src/worker/types.rs:1]: please fix clippy.

**Review by RogerKSI — CHANGES_REQUESTED** [review]

File: bothan-coinmarketcap/src/api/rest.rs [hunk 3]
```diff
@@ -22,7 +22,7 @@ impl CoinMarketCapRestAPI {
     /// Equivalent to the `/v2/cryptocurrency/quotes/latest` endpoint.
     pub async fn get_latest_quotes(
         &self,
-        ids: &[usize],
+        ids: &[&str],
```
Comment (warittornc) [inline bothan-coinmarketcap/src/api/rest.rs:25]: use usize or u32, etc

File: bothan-coinmarketcap/src/api/rest.rs [hunk 4]
```diff
@@ -22,7 +22,7 @@ impl CoinMarketCapRestAPI {
...
     ) -> Result<Vec<Option<Quote>>, RestAPIError> {
         let url = format!("{}v2/cryptocurrency/quotes/latest", self.url);
         let ids_string = ids.iter().map(|id| id.to_string()).join(",");
```
Comment (warittornc) [inline bothan-coinmarketcap/src/api/rest.rs:28]: should be able to pass `Vec<(string, usize)>` instead to params

File: bothan-coinmarketcap/src/worker/opts.rs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    DEFAULT_URL.to_string()
+}
+
+fn default_update_interval() -> Duration {
+    DEFAULT_UPDATE_INTERVAL
+}
+
+fn default_update_supported_assets_interval() -> Duration {
+    DEFAULT_UPDATE_SUPPORTED_ASSETS_INTERVAL
+}
```
Comment (warittornc) [inline bothan-coinmarketcap/src/worker/opts.rs:37]: I don't think this is still being used?

File: bothan-coinmarketcap/src/worker/opts.rs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    #[serde(default = "default_url")]
+    pub url: String,
+    #[serde(default)]
+    pub api_key: Option<String>,
+    #[serde(default = "default_update_interval")]
+    #[serde(with = "humantime_serde")]
+    pub update_interval: Duration,
+    #[serde(default = "default_update_supported_assets_interval")]
+    #[serde(with = "humantime_serde")]
+    pub update_supported_assets_interval: Duration,
```
Comment (warittornc) [inline bothan-coinmarketcap/src/worker/opts.rs:24]: I don't think this is still being used?

File: bothan-coinmarketcap/src/worker/builder.rs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        self.opts.update_interval = update_interval;
+        self
+    }
+
+    /// Sets the update supported assets interval for the `CoinMarketCapWorker`.
+    /// The default interval is `DEFAULT_UPDATE_SUPPORTED_ASSETS_INTERVAL`.
+    pub fn with_update_supported_assets_interval(mut self, update_interval: Duration) -> Self {
+        self.opts.update_supported_assets_interval = update_interval;
+        self
+    }
```
Comment (warittornc) [inline bothan-coinmarketcap/src/worker/builder.rs:49]: I don't think this is still being used

File: bothan-coinmarketcap/src/worker/opts.rs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    DEFAULT_UPDATE_SUPPORTED_ASSETS_INTERVAL
+}
+
+impl Default for CoinMarketCapWorkerBuilderOpts {
+    fn default() -> Self {
+        Self {
+            url: default_url(),
+            api_key: None,
+            update_interval: default_update_interval(),
+            update_supported_assets_interval: default_update_supported_assets_interval(),
```
Comment (warittornc) [inline bothan-coinmarketcap/src/worker/opts.rs:45]: ditto

File: bothan-coinmarketcap/src/worker/types.rs (hunk 1 above)
Comment (warittornc) [reply bothan-coinmarketcap/src/worker/types.rs:4]: ditto

**Review by warittornc — CHANGES_REQUESTED** [review]

File: bothan-coinmarketcap/src/worker/types.rs (hunk 1 above)
Comment (colmazia) [reply bothan-coinmarketcap/src/worker/types.rs]: deleted

File: bothan-coinmarketcap/src/worker/types.rs (hunk 2 above)
Comment (colmazia) [reply bothan-coinmarketcap/src/worker/types.rs:1]: fixed

File: bothan-coinmarketcap/src/api/rest.rs (hunk 3 above)
Comment (colmazia) [reply bothan-coinmarketcap/src/api/rest.rs]: fixed

File: bothan-coinmarketcap/src/worker/opts.rs (hunk 5 above)
Comment (colmazia) [reply bothan-coinmarketcap/src/worker/opts.rs]: deleted

File: bothan-coinmarketcap/src/worker/opts.rs (hunk 6 above)
Comment (colmazia) [reply bothan-coinmarketcap/src/worker/opts.rs]: deleted

File: bothan-coinmarketcap/src/worker/builder.rs (hunk 7 above)
Comment (colmazia) [reply bothan-coinmarketcap/src/worker/builder.rs]: deleted

File: bothan-coinmarketcap/src/worker/opts.rs (hunk 8 above)
Comment (colmazia) [reply bothan-coinmarketcap/src/worker/opts.rs]: deleted

**Review by RogerKSI — APPROVED** [review]

**Review by warittornc — APPROVED** [review]

**Pull request merged**